// graphql-tools-fork/dist/transforms/transformSchema.js
Object.defineProperty(exports, "__esModule", { value: true });
var makeExecutableSchema_1 = require("../makeExecutableSchema");
var transforms_1 = require("../transforms/transforms");
var resolvers_1 = require("../stitching/resolvers");
var Interfaces_1 = require("../Interfaces");
var clone_1 = require("../utils/clone");
function wrapSchema(subschemaOrSubschemaConfig, transforms) {
    var subschemaConfig = Interfaces_1.isSubschemaConfig(subschemaOrSubschemaConfig)
        ? subschemaOrSubschemaConfig
        : { schema: subschemaOrSubschemaConfig };
    var schema = clone_1.cloneSchema(subschemaConfig.schema);
    resolvers_1.stripResolvers(schema);
    makeExecutableSchema_1.addResolversToSchema({
        schema: schema,
        resolvers: resolvers_1.generateProxyingResolvers({ subschemaConfig: subschemaConfig, transforms: transforms }),
        resolverValidationOptions: {
            allowResolversNotInSchema: true,
        },
    });
    var schemaTransforms = [];
    if (subschemaConfig.transforms != null) {
        schemaTransforms = schemaTransforms.concat(subschemaConfig.transforms);
    }
    if (transforms != null) {
        schemaTransforms = schemaTransforms.concat(transforms);
    }
    return transforms_1.applySchemaTransforms(schema, schemaTransforms);
}
exports.wrapSchema = wrapSchema;
function transformSchema(subschemaOrSubschemaConfig, transforms) {
    var schema = wrapSchema(subschemaOrSubschemaConfig, transforms);
    schema.transforms = transforms.slice().reverse();
    return schema;
}
exports.default = transformSchema;
