// graphql-tools-fork/dist/polyfills/toConfig.js
// graphql <v14.2 does not support toConfig
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_1 = require("graphql");
var graphqlVersion_1 = require("../utils/graphqlVersion");
function schemaToConfig(schema) {
    if (schema.toConfig != null) {
        return schema.toConfig();
    }
    var newTypes = [];
    var types = schema.getTypeMap();
    Object.keys(types).forEach(function (typeName) {
        newTypes.push(types[typeName]);
    });
    var schemaConfig = {
        query: schema.getQueryType(),
        mutation: schema.getMutationType(),
        subscription: schema.getSubscriptionType(),
        types: newTypes,
        directives: schema.getDirectives().slice(),
        extensions: schema.extensions,
        astNode: schema.astNode,
        extensionASTNodes: schema.extensionASTNodes != null ? schema.extensionASTNodes : [],
        assumeValid: schema
            .__validationErrors !== undefined,
    };
    if (graphqlVersion_1.graphqlVersion() >= 15) {
        schemaConfig.description = schema.description;
    }
    return schemaConfig;
}
exports.schemaToConfig = schemaToConfig;
function toConfig(schemaOrTypeOrDirective) {
    if (graphql_1.isSchema(schemaOrTypeOrDirective)) {
        return schemaToConfig(schemaOrTypeOrDirective);
    }
    else if (graphql_1.isDirective(schemaOrTypeOrDirective)) {
        return directiveToConfig(schemaOrTypeOrDirective);
    }
    else if (graphql_1.isNamedType(schemaOrTypeOrDirective)) {
        return typeToConfig(schemaOrTypeOrDirective);
    }
    throw new Error("Unknown object " + schemaOrTypeOrDirective);
}
exports.toConfig = toConfig;
function typeToConfig(type) {
    if (graphql_1.isObjectType(type)) {
        return objectTypeToConfig(type);
    }
    else if (graphql_1.isInterfaceType(type)) {
        return interfaceTypeToConfig(type);
    }
    else if (graphql_1.isUnionType(type)) {
        return unionTypeToConfig(type);
    }
    else if (graphql_1.isEnumType(type)) {
        return enumTypeToConfig(type);
    }
    else if (graphql_1.isScalarType(type)) {
        return scalarTypeToConfig(type);
    }
    else if (graphql_1.isInputObjectType(type)) {
        return inputObjectTypeToConfig(type);
    }
    throw new Error("Unknown type " + type);
}
exports.typeToConfig = typeToConfig;
function objectTypeToConfig(type) {
    if (type.toConfig != null) {
        return type.toConfig();
    }
    var typeConfig = {
        name: type.name,
        description: type.description,
        interfaces: type.getInterfaces(),
        fields: fieldsToFieldsConfig(type.getFields()),
        isTypeOf: type.isTypeOf,
        extensions: type.extensions,
        astNode: type.astNode,
        extensionASTNodes: type.extensionASTNodes != null ? type.extensionASTNodes : [],
    };
    return typeConfig;
}
exports.objectTypeToConfig = objectTypeToConfig;
function interfaceTypeToConfig(type) {
    if (type.toConfig != null) {
        return type.toConfig();
    }
    var typeConfig = {
        name: type.name,
        description: type.description,
        fields: fieldsToFieldsConfig(type.getFields()),
        resolveType: type.resolveType,
        extensions: type.extensions,
        astNode: type.astNode,
        extensionASTNodes: type.extensionASTNodes != null ? type.extensionASTNodes : [],
    };
    if (graphqlVersion_1.graphqlVersion() >= 15) {
        typeConfig.interfaces = type.getInterfaces();
    }
    return typeConfig;
}
exports.interfaceTypeToConfig = interfaceTypeToConfig;
function unionTypeToConfig(type) {
    if (type.toConfig != null) {
        return type.toConfig();
    }
    var typeConfig = {
        name: type.name,
        description: type.description,
        types: type.getTypes(),
        resolveType: type.resolveType,
        extensions: type.extensions,
        astNode: type.astNode,
        extensionASTNodes: type.extensionASTNodes != null ? type.extensionASTNodes : [],
    };
    return typeConfig;
}
exports.unionTypeToConfig = unionTypeToConfig;
function enumTypeToConfig(type) {
    if (type.toConfig != null) {
        return type.toConfig();
    }
    var newValues = {};
    type.getValues().forEach(function (value) {
        newValues[value.name] = {
            description: value.description,
            value: value.value,
            deprecationReason: value.deprecationReason,
            extensions: value.extensions,
            astNode: value.astNode,
        };
    });
    var typeConfig = {
        name: type.name,
        description: type.description,
        values: newValues,
        extensions: type.extensions,
        astNode: type.astNode,
        extensionASTNodes: type.extensionASTNodes != null ? type.extensionASTNodes : [],
    };
    return typeConfig;
}
exports.enumTypeToConfig = enumTypeToConfig;
var hasOwn = Object.prototype.hasOwnProperty;
function scalarTypeToConfig(type) {
    if (type.toConfig != null) {
        return type.toConfig();
    }
    var typeConfig = {
        name: type.name,
        description: type.description,
        serialize: graphqlVersion_1.graphqlVersion() >= 14 || hasOwn.call(type, 'serialize')
            ? type.serialize
            : type._scalarConfig.serialize,
        parseValue: graphqlVersion_1.graphqlVersion() >= 14 || hasOwn.call(type, 'parseValue')
            ? type.parseValue
            : type._scalarConfig.parseValue,
        parseLiteral: graphqlVersion_1.graphqlVersion() >= 14 || hasOwn.call(type, 'parseLiteral')
            ? type.parseLiteral
            : type._scalarConfig.parseLiteral,
        extensions: type.extensions,
        astNode: type.astNode,
        extensionASTNodes: type.extensionASTNodes != null ? type.extensionASTNodes : [],
    };
    return typeConfig;
}
exports.scalarTypeToConfig = scalarTypeToConfig;
function inputObjectTypeToConfig(type) {
    if (type.toConfig != null) {
        return type.toConfig();
    }
    var newFields = {};
    var fields = type.getFields();
    Object.keys(fields).forEach(function (fieldName) {
        var field = fields[fieldName];
        newFields[fieldName] = {
            description: field.description,
            type: field.type,
            defaultValue: field.defaultValue,
            extensions: field.extensions,
            astNode: field.astNode,
        };
    });
    var typeConfig = {
        name: type.name,
        description: type.description,
        fields: newFields,
        extensions: type.extensions,
        astNode: type.astNode,
        extensionASTNodes: type.extensionASTNodes != null ? type.extensionASTNodes : [],
    };
    return typeConfig;
}
exports.inputObjectTypeToConfig = inputObjectTypeToConfig;
function directiveToConfig(directive) {
    if (directive.toConfig != null) {
        return directive.toConfig();
    }
    var directiveConfig = {
        name: directive.name,
        description: directive.description,
        locations: directive.locations,
        args: argsToArgsConfig(directive.args),
        isRepeatable: directive
            .isRepeatable,
        extensions: directive.extensions,
        astNode: directive.astNode,
    };
    return directiveConfig;
}
exports.directiveToConfig = directiveToConfig;
function fieldsToFieldsConfig(fields) {
    var newFields = {};
    Object.keys(fields).forEach(function (fieldName) {
        var field = fields[fieldName];
        newFields[fieldName] = {
            description: field.description,
            type: field.type,
            args: argsToArgsConfig(field.args),
            resolve: field.resolve,
            subscribe: field.subscribe,
            deprecationReason: field.deprecationReason,
            extensions: field.extensions,
            astNode: field.astNode,
        };
    });
    return newFields;
}
function argsToArgsConfig(args) {
    var newArguments = {};
    args.forEach(function (arg) {
        newArguments[arg.name] = {
            description: arg.description,
            type: arg.type,
            defaultValue: arg.defaultValue,
            extensions: arg.extensions,
            astNode: arg.astNode,
        };
    });
    return newArguments;
}
